src/driver-holtek/holtek8-shared.c

```diff
@@ -0,0 +1,1301 @@
+					delay = 0;
+				}
+
+				raw[event_i].release = ratbag_ev->type == RATBAG_MACRO_EVENT_KEY_RELEASED;
+				raw[event_i].delay = 1;
+
+				switch (ratbag_ev->event.key) {
+					case BTN_LEFT:   key = HOLTEK8_BUTTON_MOUSE_LEFT; break;
+					case BTN_RIGHT:  key = HOLTEK8_BUTTON_MOUSE_RIGHT; break;
+					case BTN_MIDDLE: key = HOLTEK8_BUTTON_MOUSE_MIDDLE; break;
+					case BTN_SIDE:   key = HOLTEK8_BUTTON_MOUSE_MB4; break;
+					case BTN_EXTRA:  key = HOLTEK8_BUTTON_MOUSE_MB5; break;
+					default:
+						key = ratbag_hidraw_get_keyboard_usage_from_keycode(device, ratbag_ev->event.key);
+				}
+				if (key == 0)
+					return -EINVAL;
+
+				raw[event_i].key = key;
+				event_i += 1;
+				break;
+			case RATBAG_MACRO_EVENT_WAIT:
+				delay += ratbag_ev->event.timeout;
+				break;
+			case RATBAG_MACRO_EVENT_NONE:
+				return event_i;
+			default:
+				return -EINVAL;
+		}
+	}
+
+	return event_i;
+}
+
+/*
+ * Converts raw device button data and writes to ratbag button.
+ * If the data->type is macro, reads a decoded macro from the device
+ * memory and writes to ratbag button.
+ *
+ * @return 0 on success or a negative errno.
+ */
+int
+holtek8_button_from_data(struct ratbag_button *button, struct holtek8_button_data *data)
+{
+	struct ratbag_device *device = button->profile->device;
+	const struct ratbag_button_action *action;
+	int rc;
+
+	action = holtek8_button_map_raw_to_action(data);
+
+	if (action != NULL) {
+		button->action = *action;
+		return 0;
+	}
+
+	button->action.type = RATBAG_BUTTON_ACTION_TYPE_UNKNOWN;
+
+	switch (data->type) {
+		case HOLTEK8_BUTTON_TYPE_KEYBOARD: {
+			unsigned int modifiers = data->keyboard.modifiers;
+			unsigned int key1 = ratbag_hidraw_get_keycode_from_keyboard_usage(device, data->keyboard.hid_key);
+			unsigned int key2 = ratbag_hidraw_get_keycode_from_keyboard_usage(device, data->keyboard.hid_key2);
+
+			holtek8_button_macro_new_from_keycodes(button, key1, key2, modifiers);
+			break;
+		}
+		case HOLTEK8_BUTTON_TYPE_MEDIA: {
+			uint16_t hid_code_cc = get_unaligned_le_u16(data->media.hid_key);
+			unsigned int key = ratbag_hidraw_get_keycode_from_consumer_usage(device, hid_code_cc);
+
+			holtek8_button_macro_new_from_keycodes(button, key, 0, 0);
+			break;
+		}
+		case HOLTEK8_BUTTON_TYPE_MACRO: {
+			struct holtek8_macro_event macro_events[HOLTEK8_MAX_MACRO_EVENTS] = {0};
+
+			rc = holtek8_read_macro_data(device, macro_events, data->macro.index);
+			if (rc == -EOVERFLOW)
+				return 0;
+			else if (rc < 0)
+				return rc;
+
+			holtek8_macro_from_events(button, macro_events);
+			break;
+		}
+		default:
+			log_debug(device->ratbag, "Button %u unsupported: %#x %#x %#x %#x\n", button->index, data->type, data->data[0], data->data[1], data->data[2]);
+			break;
+	}
+
+
+	return 0;
+}
+
+/*
+ * Converts ratbag button to raw device button data.
+ * If the ratbag's button action is macro, writes an encoded macro
+ * to the device memory and returns a data already pointing to the
+ * just written macro.
```

Comment:
Looks like it returns an int and writes to 'data' a copy of the written macro.
This should be like '@param[out] data Description' instead.

Reply:
Reworded a bit and described param data